File: Python/programmers/kakao/level3/표_편집.py
class Node:
    def __init__(self):
        self.removed = False
        self.prev = None
        self.next = None

def solution(n, k, cmd):
    nodeArr = [Node() for _ in range(n)]
    for i in range(1, n):
        nodeArr[i-1].next = nodeArr[i]
        nodeArr[i].prev = nodeArr[i-1]
    
    curr = nodeArr[k]
    stack = []

    for str in cmd:
        if str[0] == "U":
            x = int(str[2:])
            for _ in range(x):
                curr = curr.prev
        elif str[0] == "D":
            x = int(str[2:])
            for _ in range(x):
                curr = curr.next
        elif str[0] == "C":
            stack.append(curr)
            curr.removed = True
            up = curr.prev
            down = curr.next
            if up:
                up.next = down
            if down:
                down.prev = up
                curr = down
            else:
                curr = up
        elif str[0] == "Z":
            node = stack.pop()
            node.removed = False
            up = node.prev
            down = node.next
            if up:
                up.next = node
            if down:
                down.prev = node

    answer = ""
    for i in range(n):
        if nodeArr[i].removed:
            answer += "X"
        else:
            answer += "O"
    return answer


n, k = 8, 2
cmd = ["D 2","C","U 3","C","D 4","C","U 2","Z","Z"]	
print(solution(n, k, cmd)) # "OOOOXOOO"

n, k = 8, 2
cmd = ["D 2","C","U 3","C","D 4","C","U 2","Z","Z","U 1","C"]	
print(solution(n, k, cmd)) # "OOXOXOOO"

File: Java/programmers/kakao/level3/표편집.java
package CodeTest.Java.programmers.kakao.level3;

import java.util.*;

public class 표편집 {
    public String solution(int n, int k, String[] cmd) {
        Node[] nodeArr = new Node[n];
        for (int i=0; i<n; i++) {
            nodeArr[i] = new Node(false, null, null);
        }

        for (int i=1; i<n; i++) {
            nodeArr[i-1].next = nodeArr[i];
            nodeArr[i].prev = nodeArr[i-1];
        }

        Node cur = nodeArr[k];
        Stack<Node> stk = new Stack<>();

        for (String st : cmd) {
            char p = st.charAt(0);
            if (p == 'U') {
                int x = Integer.parseInt(st.substring(2));
                for (int i=0; i<x; i++) cur = cur.prev;
            } 
            else if (p == 'D') {
                int x = Integer.parseInt(st.substring(2));
                for (int i=0; i<x; i++) cur = cur.next;
            }
            else if (p == 'C') {
                stk.push(cur);
                cur.removed = true;
                Node up = cur.prev, down = cur.next;
                if (up != null) {
                    up.next = down;
                }
                if (down != null) {
                    down.prev = up;
                    cur = down;
                } else {
                    cur = up;
                }
            }
            else if (p == 'Z') {
                Node node = stk.pop();
                node.removed = false;
                Node up = node.prev, down = node.next;
                if (up != null) {
                    up.next = node;
                }
                if (down != null) {
                    down.prev = node;
                }
            }
        }

        StringBuilder answer = new StringBuilder();
        for (int i=0; i<n; i++) {
            if (nodeArr[i].removed) answer.append("X");
            else answer.append("O");
        }

        return answer.toString();
    }

    private class Node {
        boolean removed;
        Node prev, next;
        
        Node(boolean removed, Node prev, Node next) {
            this.removed = removed;
            this.prev = prev;
            this.next = next;
        }
    }
}
